boot/freeldr/freeldr/arch/i386/multiboot.S

```diff
@@ -98,73 +100,70 @@ MultibootEntry:
     /* Save multiboot info structure */
     mov esi, ebx
     mov edi, offset MultibootInfo + INITIAL_BASE - FREELDR_BASE
-    mov ecx, (MB_INFO_SIZE / 4)
-    rep movsd
-    mov dword ptr ds:[MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], 0
-    mov dword ptr ds:[_MultibootInfoPtr + INITIAL_BASE - FREELDR_BASE], offset MultibootInfo
+    mov ecx, MB_INFO_SIZE
+    rep movsb
+    mov dword ptr [MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], 0
+    mov dword ptr [_MultibootInfoPtr + INITIAL_BASE - FREELDR_BASE], offset MultibootInfo
 
     /* See if the memory map was passed in */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_MEMORY_MAP
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_MEMORY_MAP
     jz mbchk_command_line
     /* Check memory map length */
-    mov ecx, dword ptr ds:[ebx + MB_INFO_MMAP_LEN_OFFSET]
+    mov rcx, [rbx + MB_INFO_MMAP_LEN_OFFSET]
     test ecx, ecx
     jz mbchk_command_line
     cmp ecx, MB_MMAP_SIZE
     jg mbchk_command_line
     /* Check memory map address */
-    mov esi, dword ptr ds:[ebx + MB_INFO_MMAP_ADDR_OFFSET]
+    mov rsi, [rbx + MB_INFO_MMAP_ADDR_OFFSET]
     test esi, esi
     jz mbchk_command_line
     /* Save memory map structure */
     mov edi, offset MultibootMemoryMap + INITIAL_BASE - FREELDR_BASE
-    shr ecx, 2
-    rep movsd
+    rep movsb
     /* Relocate memory map address */
-    mov dword ptr ds:[MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], offset MultibootMemoryMap
+    mov dword ptr [MultibootInfo + INITIAL_BASE - FREELDR_BASE + MB_INFO_MMAP_ADDR_OFFSET], offset MultibootMemoryMap
 
 mbchk_command_line:
     /* Save command line */
-    test dword ptr ds:[ebx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_COMMAND_LINE
+    test dword ptr [rbx + MB_INFO_FLAGS_OFFSET], MB_INFO_FLAG_COMMAND_LINE
     jz mb2
-    mov esi, dword ptr ds:[ebx + MB_INFO_COMMAND_LINE_OFFSET]
+    mov rsi, [rbx + MB_INFO_COMMAND_LINE_OFFSET]
     mov edi, offset cmdline + INITIAL_BASE - FREELDR_BASE
     mov ecx, CMDLINE_SIZE - 1
 mb1:
     lodsb
     stosb
     test al, al
     jz mb2
-    dec ecx
+    sub ecx, 1
```

Comment:
Why this change?

Reply:
because this instruction have the same opcode for both i386 and amd64.

Reply:
And what happens if you used the dec instead?

Reply:
in 64 bits it will generate `FF C9` and that's a problem because we are in 32 bits mode.

Reply:
It would try to decrement rcx in a sense?

Reply:
No. it would decrement the ecx.